### PR TITLE
Preserve bindings metadata in block transforms

### LINKED ISSUE #59086: Metadata attribute is not preserved after transforming paragraph block into heading
## What?
It seems that when transforming a block the `metadata` attribute is not preserved. For example, if I name a paragraph block with a custom name, it disappears after transforming it into a heading. Or if it has some bindings.

It seems it works fine with other attributes.

Is this expected? If not, any idea how it could be solved?

## Reproduce

1. Add a paragraph to a post.
2. Rename the block to "Custom name".
3. Transform the paragraph block into a heading block.
4. Check that the "Custom name" has dissapeared.

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
Fixes https://github.com/WordPress/gutenberg/issues/59086

This pull request adds the possibility of preserving and adapting the metadata attribute for the following transforms:
* Paragraph to heading.
* Heading to paragraph.
* Paragraph to button.

I've also added some tests to check that, in those transformations, the content, align, and metadata attributes are preserved. I couldn't find any existing tests for that. Please let me know if that's not the case.

A couple of questions that come to my mind:
* Should we preserve any data stored in `metadata` or only the bindings?
* Should we explore the possibility of using a hook to make it work for all transforms instead of hardcoding it for these ones?

## Why?
Right now, if you rename your block, or add a binding to it, and transform it into another block, that data is not preserved.

## How?
I just added some logic in the heading, and the button transforms objects to pass the `metadata` attribute as well.

## Testing Instructions

**Test renaming a block is preserved**

* Go to a page and add a paragraph.
* Click on the block options and rename it.
* Transform paragraph into a heading.
* Check that the custom name.

Repeat the process transforming a heading into a paragraph and a paragraph into a button.

**Test bindings in a block are preserved**

* Go to a page and add a paragraph.
* Go to the Code editor and add a binding to it.
```
{"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"my_custom_field"}}}}}
```
* Go to the Visual editor and transform the paragraph into a heading.
* Check that the bindings object is preserved.

Repeat the process transforming a heading into a paragraph.

Repeat the process transforming a paragraph into a button. In this case, the object should change to point to the "text" attribute instead of "content".

### CODE REVIEW

**Comment by talldan** [comment]
Thanks for putting the PR together!

> Should we preserve any data stored in metadata or only the bindings?

IMO, the `name` could be preserved, as it feels like a bug that it's removed after a transform.

The `id` could also be preserved, as I think that would ensure the pattern overrides continue to work after transforming a block in a pattern source.

> Should we explore the possibility of using a hook to make it work for all transforms instead of hardcoding it for these ones?

The `name` definitely feels like a candidate for something that could apply universally on any transform. It also seems like a core block editor feature, so could be part of the code in the `blocks` package.

I have wondered if there's an automatic way to transform the binding attributes, for example passing in an object like `{ content: 'content' }` as the original block attributes to the transform function, and seeing what key the value 'content' get mapped to. But it gets a bit sketchy if the transform function does any manipulation to the attribute value.

Some blocks also add a new wrapper (e.g. paragraph to group/columns), so there are some interesting permutations. It might be difficult to handle this completely automatically.

The maintenance of handling every block individually is also quite a lot to consider though 😄

File: test/e2e/specs/editor/various/convert-block-type.spec.js
Comment (gziolo) [inline test/e2e/specs/editor/various/convert-block-type.spec.js]: There are existing tests for individual blocks that verify transforms, so maybe it would be better to move them there as there might some duplication created over time:

Screenshot 2024-02-20 at 09 56 01

File: packages/block-library/src/buttons/transforms.js
Comment (gziolo) [inline packages/block-library/src/buttons/transforms.js]: It feels like the approach used for attributes would work better also for metadata for now. In effect, I would compare the current implementation with the alternative, where metadata is explicitly transformed based on the shape we expect:

```js
return createBlock( 'core/button', {
	text,
	url,
	metadata: {
		id: metadata?.id,
		name: metadata?.name,
		bindings: {
			text: metadata?.bindings?.content
		}
	},
} );
```

We would need a more sophisticated util, that would only create metadata if it contains any data, plus it would have to only include keys that are defined.

**Comment by gziolo** [comment]
First of all, major props to @SantosGuillamot for proposing the initial implementation so spark the discussion.

> The maintenance of handling every block individually is also quite a lot to consider though 😄

The same issue exists for block attributes. Since block bindings depend on them, I don't think we can automate in any way without doing the same for the attributes themselves.

> IMO, the `name` could be preserved, as it feels like a bug that it's removed after a transform.

> The `id` could also be preserved, as I think that would ensure the pattern overrides continue to work after transforming a block in a pattern source.

I think the biggest challenge is that `metadata` is also an attribute and while we can automatically preserve some of them, the question is where is the line. For example, what happens when you transform the Paragraph block into the Verse block where today we don't support block bindings? 

In effect, I would be in favor of making the transformation explicit for now and make it coded in the cases where it makes sense today. In the future, if we repeat something in all places, we can abstract it away.

File: test/e2e/specs/editor/various/convert-block-type.spec.js
Comment (SantosGuillamot) [reply test/e2e/specs/editor/various/convert-block-type.spec.js]: I moved the transform tests to the individual blocks in [this commit](https://github.com/WordPress/gutenberg/pull/59179/commits/a6fa4ce8623caf8a1541fcc6db35b6ba6f1d49eb).

File: packages/block-library/src/buttons/transforms.js
Comment (SantosGuillamot) [reply packages/block-library/src/buttons/transforms.js]: I've changed the implementation of the button transform in this commit: [link](https://github.com/WordPress/gutenberg/pull/59179/commits/51689629dce80e92aa455a6c88207ad8b0202650). Basically, that way:
* It only adds the supported props (id, name, and bindings) to the resulting button.
* It only adds the props that are included in the original paragraph metadata.
* It transforms the paragraph `content` into the button `text`.
* If the original paragraph block doesn't have `metadata`, it is not included in the resulting button.

Let me know if something like that is what you had in mind. Would it make sense to do something similar for the heading?

Comment (gziolo) [reply packages/block-library/src/buttons/transforms.js]: Yes, it looks great. Can we have the same approach for other block transforms updated?

**Comment by gziolo** [comment]
This is looking great for the initial rollout after the iterations applied from my perspective. I left one https://github.com/WordPress/gutenberg/pull/59179#discussion_r1495970444 regarding replicating the approach fro the Button block on other modified blocks.

File: packages/block-library/src/buttons/transforms.js
Comment (SantosGuillamot) [reply packages/block-library/src/buttons/transforms.js]: I used a similar logic for the heading block in [this commit](https://github.com/WordPress/gutenberg/pull/59179/commits/e472ebfc0d16b234bb4a567c057a8e330fdd8eef). With that, the block transform impacted by block bindings are covered:
* Paragraph to heading.
* Heading to paragraph.
* Paragraph to button.

I can take a look at other block transforms that might want to preserve the id or the name and explore if I can create a util. Any clue where that util should live?

Comment (SantosGuillamot) [reply packages/block-library/src/buttons/transforms.js]: This is an initial idea of how the util could work: [link](https://github.com/WordPress/gutenberg/pull/59179/commits/bb3b32565f9d1ea8009dabcd2632e65935a434b4).

I also added another example with the Code block to see how we could make `name` and `id` persist, not only for blocks that are supported by block bindings: [link](https://github.com/WordPress/gutenberg/pull/59179/commits/7eb0fdb60ecf4f1cfe2951d4c96da4aa018e666e).

Let me know if I should keep pursuing that path or I should just revert it and keep the scope limited to the block bindings. If this makes sense, I can take care of reviewing the block transforms and use the util there to maintain the `name` and `id`.

**Comment by artemiomorales** [comment]
I gave this a test and it's all working as expected for me, and nothing jumps out at me regarding the util logic 👍

I've tried rerunning the [failing mobile tests](https://github.com/WordPress/gutenberg/actions/runs/7977186378/job/21796752274?pr=59179) and they failed again — I'm not sure why `getTransformedMetadata` is not being found.

**Comment by gziolo** [comment]
> I've tried rerunning the [failing mobile tests](https://github.com/WordPress/gutenberg/actions/runs/7977186378/job/21796752274?pr=59179) and they failed again — I'm not sure why `getTransformedMetadata` is not being found.

Good catch! It looks like the code for Gutenberg Mobile has some overridden files that don't return the new helper method. My recommendation would be to move the util to `packages/block-library/src/utils/` folder. This way the file is going to be available to the React Native code without any additional work. Additionally, there isn't much value at the moment in locating it inside `@wordpress/block-editor` since it's private and only used by core blocks.

File: packages/block-library/src/buttons/transforms.js
Comment (gziolo) [reply packages/block-library/src/buttons/transforms.js]: Yes, that looks like you are on a good path. It should make it easier to add handling for `attributes.metadata` to all existing transforms between blocks. We can revisit the approach and see it there is room for improvement once the issue is fixed on time for WP 6.5 Beta 3.

**Comment by SantosGuillamot** [comment]
> My recommendation would be to move the util to packages/block-library/src/utils/ folder. 

This makes sense to me. I believe mobile tests are failing because I didn't expose the function in [the private-apis.native.js file](https://github.com/WordPress/gutenberg/blob/fix/preserve-metadata-in-block-transforms/packages/block-editor/src/private-apis.native.js).

EDIT: I changed it in [this commit](https://github.com/WordPress/gutenberg/pull/59179/commits/b2d54b20a586fc1546e2299320112d74ebb3fc9d).

**Review by gziolo — APPROVED** [review]
Nice one 👍🏻 

Let's make sure CI is back to normal and all tests pass.

**Comment by SantosGuillamot** [comment]
In [this commit](https://github.com/WordPress/gutenberg/pull/59179/commits/d32350783f828527a49b1cbf27460c920f679870#diff-b54a832f17bbe276f0872557dc17de4a2010c2c343186b82fe384500b25fbb1e), I changed the util trying to understand which properties should be preserved based on the block supports. The goal was to make it easier to use. This way, you just have to pass the name of the original and the resulting blocks.

I believe this should help to abstract it further in the future. Let me know what you think.

I'll add some tests to the code block as well.

EDIT: Code block transform tests added.

File: packages/block-library/src/utils/get-transformed-metadata.js
Comment (gziolo) [inline packages/block-library/src/utils/get-transformed-metadata.js]: If `transformSupportedProps` is empty, it's possible to return early.

**Comment by SantosGuillamot** [comment]
I believe I implemented all the changes I had in mind. I will work on supporting block renaming in other block transform in a follow-up PR.

File: packages/block-library/src/utils/get-transformed-metadata.js
Comment (gziolo) [inline packages/block-library/src/utils/get-transformed-metadata.js]: It's probably enough to pass an optional second param that is an object with mapping between bindings:

```js
getTransformedMetadata( metadata, { content: 'text' } );
```

or with a callback that would be more flexible as you could pass any attribute:

```js
getTransformedMetadata( metadata, ( { bindings: { content } } ) => ( { bindings: { text: content } } );
```

File: packages/block-library/src/utils/get-transformed-metadata.js
Comment (SantosGuillamot) [reply packages/block-library/src/utils/get-transformed-metadata.js]: You're right. Done in [this commit](https://github.com/WordPress/gutenberg/pull/59179/commits/2d09640e3bee54fdb8893e8275d852c9a7152087).

File: packages/block-library/src/utils/get-transformed-metadata.js
Comment (SantosGuillamot) [reply packages/block-library/src/utils/get-transformed-metadata.js]: This makes sense 🙂 

I added the callback in [this commit](https://github.com/WordPress/gutenberg/pull/59179/commits/5e3eda0efe3b2b2cedd2aafe078ef0a6427e6779). I preserved the `toBlockName` property because it is needed to get the block supports.

Comment (gziolo) [reply packages/block-library/src/utils/get-transformed-metadata.js]: All good 👍🏻

**Review by gziolo — APPROVED** [review]

**Comment by ntsekouras** [comment]
Thanks for the PR! This is an interesting challenge here and this seems like a first iteration that actually kind of centralizes the handling, even not being a generic solution.  I'm definitely going to check it more when I'm back(currently in a team meetup).

**Comment by SantosGuillamot** [comment]
Thanks for jumping in 🙂 

>I'm definitely going to check it more when I'm back(currently in a team meetup).

Would you like us to wait until you review it or should we merge it and work on improvements in follow-up PRs?

**Comment by ntsekouras** [comment]
> Would you like us to wait until you review it 

I have checked the code and tested already, and others have reviewed. Of course you can land :) Great job here @SantosGuillamot !

**Comment by getdave** [comment]
I just cherry-picked this PR to the cherry-pick-wp-6-5-beta-3 branch to get it included in the next release: 6a8bf9d28e

**Pull request merged**
